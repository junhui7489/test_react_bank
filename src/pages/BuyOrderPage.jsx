import React, {useContext, useState} from 'react';
import { OrderTradesContext} from '../contexts/OrderTradesContext';
import { AccountBalanceContext } from '../contexts/AccountBalanceContext';
import ButtonAppBar from './components/Appbar';
import BuyOrdersTable from './components/BuyOrdersTable';
import AddMoneyButton from './components/AddMoneyButton';

export default function BuyOrderPage(props){
  const [balance, setBalance] = useContext(AccountBalanceContext);
  const [allTrades, getTradesData] = useContext(OrderTradesContext); 

  return(
    <div>
    <ButtonAppBar/>
    <br/>
    <span className='d-flex justify-content-between'>
    <h3>My Account Balance is ${balance}.</h3>
    <AddMoneyButton label="Add Funds"/>
    </span>
    <br/>
    <BuyOrdersTable data={allTrades}/>
    </div>
  )
}