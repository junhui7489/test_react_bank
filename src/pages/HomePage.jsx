import React, {useContext, useEffect, useState} from 'react';
import { PersonalOrderContext } from '../contexts/PersonalOrderContext';
import { AccountBalanceContext } from '../contexts/AccountBalanceContext';
import ButtonAppBar from './components/Appbar';
import PersonalOrdersTable from './components/PersonalOrdersTable';


export default function HomePage(props){

  const [listDetails, setListDetails] = useContext(PersonalOrderContext);
  const [balance, setBalance] = useContext(AccountBalanceContext);

  return(
    <div>
      <ButtonAppBar/>
      <br/>
      <span>
      <h3>My Account Balance is ${balance}.</h3>
      </span>
      <br/>
      <PersonalOrdersTable data={listDetails}/>
    </div>
  )
}