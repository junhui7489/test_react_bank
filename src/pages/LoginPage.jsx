import React, { Component } from "react";
import "../App.css";
import logo from "../assets/logo-social.png";
import { Link, useNavigate } from "react-router-dom";
import { ToastContainer, toast } from 'react-toastify'; 
import 'react-toastify/dist/ReactToastify.css';

function LoginPage(){

    let navigate = useNavigate();

    function handleSubmit(e) {
      try {
        e.preventDefault();    
        let payload = {
          "username": e.target.username.value,
          "password": e.target.password.value
        };

        let final_payload = JSON.stringify(payload);

        if (!e.target.username.value) {
          toast.error("Username is required")
          throw new Error("username is required");
        } else if (!e.target.password.value) {
          toast.error("Password is required")
          throw new Error("Password is required");
        } else{
          fetch("http://localhost:3000/api/login",
          {
            method: "post",
            headers: {'Content-Type': 'application/json'},
            body: final_payload
          }).then(
            (res) =>{
              console.log(res);
              let response = res.json();
              console.log(response);
              response.then((data)=>{
                console.log(response);
                if (data.hasOwnProperty('message')) {
                  toast.success("Successfully logged in, John!");
                  setTimeout(()=>{navigate('/home')}, 2500);
                  e.target.username.value = "";
                  e.target.password.value = "";
                }
                else{
                  toast.error("Wrong email or password. Please try again!");
                  throw new Error("Wrong email or password combination");
                }
              })

            });
          } 
          } catch (error) {
            console.log(error);
          }
      };
    
        return (
          <div className="wrapper">
            <ToastContainer/>
            <img src={logo} className="logo" alt="Logo" />
            <br/>
            <div className="col text-center">
              <h2 className="display-5">Login</h2>
            </div>
            <form className="form" onSubmit={handleSubmit} noValidate>
              <div className="mb-3">
                <label className="form-label" htmlFor="username">Username</label>
                <input type="text" class="form-control" name="username" placeholder="Username" required aria-required="true" />
              </div>
              <div className="mb-3">
                <label className="form-label" htmlFor="password">Password</label>
                <input type="password" class="form-control" name="password" placeholder="password" required aria-required="true"/>
              </div>
              <br/>
              <div className="col text-center">
              <button className="btn btn-primary mb-3 ">ENTER</button>
              </div>
            </form>
          </div>
        );
}

export default LoginPage;