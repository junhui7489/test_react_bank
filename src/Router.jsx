import React from 'react';
import { Routes, Route, BrowserRouter} from 'react-router-dom';
import LoginPage from './pages/LoginPage';
// import ForgotPasswordPage from './pages/ForgotPasswordPage'
import HomePage from './pages/HomePage';
import ErrorBoundary from './ErrorBoundary';
import BuyOrderPage from './pages/BuyOrderPage';


const Router = () => (
    <ErrorBoundary>
    <BrowserRouter>
        <Routes>
            <Route path="/" element={<LoginPage />} />
            <Route path="/login" element={<LoginPage />} />
            <Route path="/home" element={<HomePage />} />
            <Route path="/buy_order" element={<BuyOrderPage />} />
        </Routes>
    </BrowserRouter>
    </ErrorBoundary>
);

export default Router;