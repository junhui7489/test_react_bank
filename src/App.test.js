import { render, screen } from '@testing-library/react';
import LoginPage from './pages/LoginPage';
import { MemoryRouter } from 'react-router-dom';
import Router from './Router';
import PersonalOrderProvider from './contexts/PersonalOrderContext';
import AccountBalanceProvider from './contexts/AccountBalanceContext';
import OrderTradesProvider from './contexts/OrderTradesContext';
import HomePage from './pages/HomePage';
import MockPersonalData from './test/test_personal_mock.json';
import BuyOrderPage from './pages/BuyOrderPage';

describe("Login Page", ()=>{
  test('Login text', ()=> {
    render(
      <LoginPage/>, {wrapper: MemoryRouter}
    )
    const linkElement = screen.getByText(/Login/);
    expect(linkElement).toBeInTheDocument();
  })

  test('Login snapshot text', ()=> {
    render(
    <LoginPage/>, {wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText(/Login/);
    expect(linkElement).toMatchSnapshot();
  })

  test('Username text', ()=> {
    render(
      <LoginPage/>, {wrapper: MemoryRouter}
    )
    const linkElement = screen.getByText("Username");
    expect(linkElement).toBeInTheDocument();
  })

  test('Username snapshot text', ()=> {
    render(
    <LoginPage/>, {wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText("Username");
    expect(linkElement).toMatchSnapshot();
  })

  test('Password text', ()=> {
    render(
      <LoginPage/>, {wrapper: MemoryRouter}
    )
    const linkElement = screen.getByText("Password");
    expect(linkElement).toBeInTheDocument();
  })

  test('Password snapshot text', ()=> {
    render(
    <LoginPage/>, {wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText("Password");
    expect(linkElement).toMatchSnapshot();
  })

})


describe("Personal Order Page tests", ()=>{
  test('personal order summary text', ()=>{
    render(
      <AccountBalanceProvider>
        <PersonalOrderProvider>
        <HomePage/>
        </PersonalOrderProvider>
      </AccountBalanceProvider>,{wrapper: MemoryRouter}
    )
    const linkElement = screen.getByText("Personal Order Summary");
    expect(linkElement).toBeInTheDocument();
  })

  

  test('buy trades text', ()=>{
    render(
      <PersonalOrderProvider>
        <AccountBalanceProvider>
            <HomePage/>
        </AccountBalanceProvider>
      </PersonalOrderProvider>, {wrapper: MemoryRouter}
    )
    const linkElement = screen.getByText(/Buy Trade/);
    expect(linkElement).toBeInTheDocument();

  })

  test('logout text', ()=>{
    render(
      <PersonalOrderProvider>
        <AccountBalanceProvider>
            <HomePage/>
        </AccountBalanceProvider>
      </PersonalOrderProvider>, {wrapper: MemoryRouter}
    )
    const linkElement = screen.getByText("Logout")
    expect(linkElement).toBeInTheDocument();

  })

  test('Table contains company name label', () => {
    render(
      <PersonalOrderProvider>
        <AccountBalanceProvider>
            <HomePage/>
        </AccountBalanceProvider>
      </PersonalOrderProvider>, {wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText("Company Name");
    expect(linkElement).toBeInTheDocument();
  });

  test('Table contains company type label', () => {
    render(
      <PersonalOrderProvider>
        <AccountBalanceProvider>
            <HomePage/>
        </AccountBalanceProvider>
      </PersonalOrderProvider>, {wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText("Company Type");
    expect(linkElement).toBeInTheDocument();
  });

  test('Table contains market cap label', () => {
    render(
      <PersonalOrderProvider>
        <AccountBalanceProvider>
            <HomePage/>
        </AccountBalanceProvider>
      </PersonalOrderProvider>, {wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText("Market Cap");
    expect(linkElement).toBeInTheDocument();
  });

  test('Table contains current market price label', () => {
    render(
      <PersonalOrderProvider>
        <AccountBalanceProvider>
            <HomePage/>
        </AccountBalanceProvider>
      </PersonalOrderProvider>, {wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText("Current Market Price (USD)");
    expect(linkElement).toBeInTheDocument();
  });

  test('Table contains cost price label', () => {
    render(
      <PersonalOrderProvider>
        <AccountBalanceProvider>
            <HomePage/>
        </AccountBalanceProvider>
      </PersonalOrderProvider>, {wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText("Cost Price (USD)");
    expect(linkElement).toBeInTheDocument();
  });

  test('Table contains number of units label', () => {
    render(
      <PersonalOrderProvider>
        <AccountBalanceProvider>
            <HomePage/>
        </AccountBalanceProvider>
      </PersonalOrderProvider>, {wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText("Number of units");
    expect(linkElement).toBeInTheDocument();
  });

  test('Table contains bought date label', () => {
    render(
      <PersonalOrderProvider>
        <AccountBalanceProvider>
            <HomePage/>
        </AccountBalanceProvider>
      </PersonalOrderProvider>, {wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText(/Bought Date/);
    expect(linkElement).toBeInTheDocument();
  });

})

describe("Buy Order Page tests", ()=>{
  test('personal order summary text', ()=>{
    render(
      <AccountBalanceProvider>
        <OrderTradesProvider>
        <BuyOrderPage/>
        </OrderTradesProvider>
      </AccountBalanceProvider>,{wrapper: MemoryRouter}
    )
    const linkElement = screen.getByText("Personal Order Summary");
    expect(linkElement).toBeInTheDocument();
  })

  test('buy trades text', ()=>{
    render(
      <AccountBalanceProvider>
        <OrderTradesProvider>
        <BuyOrderPage/>
        </OrderTradesProvider>
      </AccountBalanceProvider>,{wrapper: MemoryRouter}
    )
    const linkElement = screen.getByText(/Buy Trade/);
    expect(linkElement).toBeInTheDocument();

  })

  test('logout text', ()=>{
    render(
      <AccountBalanceProvider>
        <OrderTradesProvider>
        <BuyOrderPage/>
        </OrderTradesProvider>
      </AccountBalanceProvider>,{wrapper: MemoryRouter}
    )
    const linkElement = screen.getByText("Logout")
    expect(linkElement).toBeInTheDocument();

  })

  test('Table contains company name label', () => {
    render(
      <AccountBalanceProvider>
        <OrderTradesProvider>
        <BuyOrderPage/>
        </OrderTradesProvider>
      </AccountBalanceProvider>,{wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText("Company Name");
    expect(linkElement).toBeInTheDocument();
  });

  test('Table contains Current Market Price label', () => {
    render(
      <AccountBalanceProvider>
        <OrderTradesProvider>
        <BuyOrderPage/>
        </OrderTradesProvider>
      </AccountBalanceProvider>,{wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText("Current Market Price (USD)");
    expect(linkElement).toBeInTheDocument();
  });

  test('Table contains `Available for Allocation` label', () => {
    render(
      <AccountBalanceProvider>
        <OrderTradesProvider>
        <BuyOrderPage/>
        </OrderTradesProvider>
      </AccountBalanceProvider>,{wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText("Available for Allocation");
    expect(linkElement).toBeInTheDocument();
  });

  test('Table contains action label', () => {
    render(
      <AccountBalanceProvider>
        <OrderTradesProvider>
        <BuyOrderPage/>
        </OrderTradesProvider>
      </AccountBalanceProvider>,{wrapper: MemoryRouter}
    );
    const linkElement = screen.getByText("Action");
    expect(linkElement).toBeInTheDocument();
  });

})

describe("test data", ()=>{
  test('Json length of at least 10', ()=>{
    expect(MockPersonalData.length).toBeGreaterThanOrEqual(2);
  })

  test('Json contains correct keys', ()=>{
    for(let i of MockPersonalData) {
      expect(Object.keys(i)).toEqual(["id", "currency", "displayName", 
      "marketCap", "marketState", "messageBoardId", "postMarketChange", 
      "quoteType", "region", "regularMarketChange", 
      "regularMarketChangePercent", "regularMarketDayHigh", 
      "regularMarketDayLow", "regularMarketPrice", 
      "userCostPrice", "units", "userBuyDate"])
    }  
  })

  test('Display Name must be string', () =>{
    for(let i of MockPersonalData) {
      expect(typeof i['displayName']).toBe('string');
    }  
  })
})


