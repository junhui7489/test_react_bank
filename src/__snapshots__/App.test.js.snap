// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Login snapshot text 1`] = `
<h2
  class="display-5"
>
  Login
</h2>
`;

exports[`Login Page Password snapshot text 1`] = `
<label
  class="form-label"
  for="password"
>
  Password
</label>
`;

exports[`Login Page Username snapshot text 1`] = `
<label
  class="form-label"
  for="username"
>
  Username
</label>
`;
